### ISSUE TITLE
Add ability to keep session open in case of a worker restart.

### ISSUE THREAD

**Issue by mfateev** [temporalio/sdk-go#937]
**Is your feature request related to a problem? Please describe.**
When a worker that owns a session is restarted, the session fails and returns an error to the workflow that created the session.
This creates a problem. For many use cases, a restart of a worker is not a problem as the shared state is cached on the host disk, and the session should continue as if nothing happened. For these use cases, the session should fail only if the worker is down for a while, which indicates the host failure.

**Describe the solution you'd like**
Support special worker id that is durable across restarts and can be used to re-establish a session.
Introduce a timeout that would support automatic session reestablishing in case of worker restarts.

**Comment by AhmedMozaly** [comment]
The current session implementation assumes that the state is temporarily preserved in the worker. Causing the session to fail when the worker restart.

In real-life scenarios, I think the common usage of sessions is just to route workflow activities to the same worker. As the state is mostly persisted on worker disk. So, I would vote for splitting the session concept into two child concepts:

* `Session` which keeps the currently implemented behavior
* `DurableSession` Should NOT fail on worker restart. just retry activities on the same worker. (I claim that this is the more common usage for sessions)

**Comment by F1bonacc1** [comment]
Hi,

As part of my workflow in kubernetes, I need to drain all the nodes one by one.
There is no way to avoid the worker pod being restarted.
Is there any update on this issue?

**Comment by Quinn-With-Two-Ns** [comment]
Hi @F1bonacc1 ,

There is no update on this issue. The best way to achieve similar behaviour is to create a task queue per worker. Then you assign consistent task queue across restarts.

We show how to create per task queue workers [here](https://github.com/temporalio/samples-go/tree/main/activities-sticky-queues)

**Comment by AhmedMozaly** [comment]
Using `a task queue per worker` won't be the best solution when `retry on another worker if the current one dies`
What we came up with. Is to:
* Each worker would listen on two task queues `public-queue` & `unique-queue-per-worker` 
* Use a session for the whole workflow
* The first activity in the workflow would choose the right task queue to use for all following activities 
* If the worker dies, the whole workflow would be retried
* When a retries happens, the first activity would choose a task queue for a healthy worker

**Comment by mohamedazouz** [comment]
@AhmedMozaly I think this example of [session-failure](https://github.com/temporalio/samples-go/tree/main/session-failure) by @Quinn-With-Two-Ns, is much much easier and straight forward than having a workaround task queues.

**Comment by AhmedMozaly** [comment]
@mohamedazouz [session-failure](https://github.com/temporalio/samples-go/tree/main/session-failure) by @Quinn-With-Two-Ns, is easier for simpler requirements.

We had a scenario with more requirements:
*  if the worker is restarted (for example due to new version release) we need to fail the session but retry **on the same** worker
* if the worker dies, fail the session on any other worker

It would be great to see this scenario has built-in support in Temporal

**Comment by mohamedazouz** [comment]
@AhmedMozaly I agree, I think this should come as native functionality when a session failed should be restarted as a whole in new or previous workers based on options
